src/internet_identity/src/main.rs

```diff
@@ -221,19 +222,37 @@ fn post_upgrade(maybe_arg: Option<InternetIdentityInit>) {
                 ));
             }
         }
-        if let Some(archive_hash) = arg.archive_module_hash {
-            state::persistent_state_mut(|persistent_state| {
-                persistent_state.archive_info.expected_module_hash = Some(archive_hash);
-            })
+        if let Some(new_config) = arg.archive_config {
+            update_archive_config(new_config);
         }
         if let Some(cost) = arg.canister_creation_cycles_cost {
             state::persistent_state_mut(|persistent_state| {
                 persistent_state.canister_creation_cycles_cost = cost;
             })
         }
+        if let Some(flag) = arg.upgrade_persistent_state {
+            if flag {
```

Comment:
beware of boolean blindness with this naked boolean; also probably simpler to match on `Some(true)` directly